docs/designs/data_transform.md

```diff
@@ -1,30 +1,79 @@
+Let's take [normalizing(min-max normalization)](https://en.wikipedia.org/wiki/Feature_scaling) the column value `capital_gain` in [census income dataset](https://archive.ics.uci.edu/ml/datasets/Census+Income) for example. The following is the `preprocess_fn` definition with TensorFlow Transform:
+
+```python
+import tensorflow_transform as tft
+
+def preprocess_fn(inputs):
+    outputs = inputs.copy()
+    outputs["capital_gain"] = tft.scale_to_0_1(inputs["capital_gain"])
+    return outputs
+```
+
+From users' perspective, SQLFlow users prefer to write SQL instead of python. It's user unfriendly if we integrate TF Transform with SQLFlow directly.  
+
+### SQL
+
+User can write SQL to do the analysis and transform work with [built-in functions](https://www.alibabacloud.com/help/doc-detail/96342.htm?spm=a2c63.p38356.b99.111.27e27309rgC5m1) or [UDF(User Defined Function)](https://www.alibabacloud.com/blog/udf-development-guide-with-maxcompute-studio_594738). The following is the SQL expression to normalize the column `capital_gain`.  
+
+```SQL
+SELECT 1.00 * (t1.capital_gain - t2.capital_gain_min) / t2.capital_gain_range
+FROM census_income t1
+JOIN
+(
+    SELECT
+        MIN(capital_gain) AS capital_gain_min,
+        MAX(capital_gain) - MIN(capital_gain) AS capital_gain_range
+    FROM census_income
+) t2
+ON 1 = 1
+```
+
+But SQL and UDF are only suitable for batch processing, we can't use SQL to transform the data at real time for serving. We need reimplement the same transform logic with other programming languages in the model serving engine. It may bring the training/serving skew.  
+
+### Internal System
+
+The feature engineering library in the internal system is configuration driven. It contains some primitive transform ops and users compose the transform logic with configuration file. The inference engine is compiled with the internal library. Users need both feature engineering configuration files and SavedModel to deploy the model for serving.  
+
+## Our Approach
+
+From the perspective of SQLFLow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of transform work as transformer. For dense column, we can use [numeric_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/numeric_column) and pass a user defined function *normalizer_fn* to convert the column value. For sparse column, we can use [embedding_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column) to map the sparse value to embedding vector or use [cross_column](https://www.tensorflow.org/api_docs/python/tf/feature_column/crossed_column) to do the feature crossing. We plan to use SQL and feature column/keras preprocessing layer together to do the data transform work.  
```

Comment:
transform work -> the transform work

Reply:
to embedding vector -> to an embedding vector

Reply:
Good suggestion, updated.